server/src/main/java/org/elasticsearch/index/mapper/DynamicTemplate.java

```diff
@@ -255,19 +347,26 @@ private static void addEntriesToPatternList(List<String> matchList, String propN
     }
 
     private static void validatePatterns(String templateName, MatchType matchType, List<String> patterns) {
-        for (String regex : patterns) {
+        for (String pattern : patterns) {
             try {
-                matchType.matches(regex, "");
+                matchType.isValidPattern(pattern);
             } catch (IllegalArgumentException e) {
-                throw new MapperParsingException(
-                    Strings.format(
-                        "Pattern [%s] of type [%s] is invalid. Cannot create dynamic template [%s].",
-                        regex,
-                        matchType,
-                        templateName
-                    ),
-                    e
-                );
+                if (matchType == MatchType.DEFAULT) {
```

Comment:
Can we move this logic into `isValidPattern` directly, rather than throwing and catching exceptions?

Reply:
So the reason I think this fits better in `validate` is that there are two branches inside this catch clause, and each only applies to one match type.  There's no shared code here.  In other words, these are bits of MatchType-specific logic, which to me means that they should live within the different MatchType implementations.

Reply:
Sounds good. I'll move it.

Reply:
This change required more code changes that I expected. Key changes:

1) I added templateName to the `MatchType.validate` params list, because the error message (that was already in place before I started this series of tickets) requires the template name in the error message.

2) Testing `MatchType.validate` with only side effects and no return value became problematic, so I changed it to return a boolean (useful mostly for unit tests) and thus changed the name back to `isValidPattern`.

Reply:
Actually, nevermind the second comment above. After consideration I found that testing without a boolean return value is straightforward. Pushed new commit.